src/main/java/software/amazon/smithy/lsp/ext/SmithyProject.java

```diff
@@ -0,0 +1,209 @@
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import software.amazon.smithy.lsp.SmithyInterface;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.SourceLocation;
+import software.amazon.smithy.model.shapes.ShapeType;
+import software.amazon.smithy.model.validation.ValidatedResult;
+
+public final class SmithyProject {
+    private List<Path> imports;
+    private List<File> smithyFiles;
+    private List<File> externalJars;
+    private Map<String, List<Location>> locations = Collections.emptyMap();
+    private ValidatedResult<Model> model;
+    private File root;
+
+    private SmithyProject(List<Path> imports, List<File> smithyFiles, List<File> externalJars, File root,
+                          ValidatedResult<Model> model) {
+        this.imports = imports;
+        this.root = root;
+        this.model = model;
+        this.smithyFiles = smithyFiles;
+        this.externalJars = externalJars;
+        model.getResult().ifPresent(m -> {
+            this.locations = collectLocations(m);
+        });
+    }
+
+    /**
+     * Recompile the model, optionally adding a file to the tracked list of sources.
+     *
+     * @param file file which may or may not be already tracked by this project
+     * @return either an error, or a loaded project
+     */
+    public Either<Exception, SmithyProject> recompile(File file) {
+        // We aggressively re-build the model with only existing files
+        // it's simpler than trying to manage which file was added/removed/closed etc.
+        List<File> newFiles = new ArrayList<File>();
+        newFiles.addAll(onlyExistingFiles(this.smithyFiles));
+
+        if (file.isFile()) {
+            newFiles.add(file);
+        }
+
+        return load(this.imports, this.smithyFiles, this.externalJars, this.root);
+    }
+
+    public ValidatedResult<Model> getModel() {
+        return this.model;
+    }
+
+    public List<File> getExternalJars() {
+        return this.externalJars;
+    }
+
+    public List<File> getSmithyFiles() {
+        return this.smithyFiles;
+    }
+
+    public Map<String, List<Location>> getLocations() {
+        return this.locations;
+    }
+
+    public Either<Exception, SmithyProject> reload(SmithyBuildExtensions config) {
+        return load(config, this.root);
+    }
+
+    /**
+     * Load the project using a SmithyBuildExtensions configuration and workspace
+     * root.
+     *
+     * @param config configuration
+     * @param root   workspace root
+     * @return either an error or a loaded project
+     */
+    public static Either<Exception, SmithyProject> load(SmithyBuildExtensions config, File root) {
+        List<Path> imports = config.getImports().stream().map(p -> Paths.get(root.getAbsolutePath(), p).normalize())
+                .collect(Collectors.toList());
+
+        if (!imports.contains(Paths.get("."))) {
+            imports.add(root.toPath());
+        }
```

Comment:
We implicitly assume all `.smithy` files in workspace root are to be added to root project.

Reply:
Yeah that's cool for now